lib/typhoeus/normalized_header_hash.rb

```diff
@@ -1,5 +1,28 @@
 module Typhoeus
   class NormalizedHeaderHash < ::Hash
+    MAX_CACHED = 64
+
+    # inspired by https://github.com/discourse/mini_mime
+    class Cache
+      def initialize(size)
+        @size = size
+        @hash = {}
+      end
+
+      def []=(key, val)
+        @hash[key] = val
+        @hash.shift if @hash.length > @size
```

Comment:
(I'm a bit late to the party, I know):
ruby's hash implementation does not guarantee which key, value pair is removed by `shift` (related to: do not rely on the order in hashes, they're an impl detail and do not belong to the `hash contract`)
in other words, the `shift` can potentially remove the just added key, val, instead of e.g. the LRU one.

Reply:
(there is an LRU cache from Sam Saffron btw)

Reply:
> do not rely on the order in hashes, they're an impl detail and do not belong to the hash contract

I don't think this is true. Since 1.9 it is documented that the insertion order is kept (see http://ruby-doc.org/core-2.6.3/Hash.html): 
> Hashes enumerate their values in the order that the corresponding keys were inserted.

But the `Hash#shift` docs indeed just say that it `removes a key-value pair from hsh` and does not mention that this will be the first one (http://ruby-doc.org/core-2.6.3/Hash.html#method-i-shift), @serioushaircut has a point.

Reply:
thx for clarifying @boosty. 
If the order is in the contract then I conclude `shift` does it correctly (removes oldest) even if the doc does not say so (found at least one blogpost that explicitly states that `shift` worx like this (I didn't dig into the ruby source). 
If unsure, another possibility under the prerequisite 'order is in the contract: `@hash.delete(@hash.keys[-1])` 
But, yeah that's ugly, at least it relies only on the key ordering.

Reply:
The implementation of Hash#shift removes the entry inserted first. It's stated in the source code. The implementation here is therefore completely fine. However, I would reverse the shift and the insertion operation. Additionally, it should be either a `while >=` or `if ==`.